### DISCUSSION TITLE
How to handle responsive GUIs

### DISCUSSION THREAD

**Discussion by 98teg** [godotengine/godot-proposals#4127] [GUI]
I'm currently working on an app that can scale accordingly to the window size. For example, I want to create a label that would automatically reduce the displayed text if the space available is not enough. Similar to ellipsis in CSS. The scene hierarchy is the following:

- PanelContainer (Full Rect)
  - VBoxContainer
    - Label

I would then make the Label listen to the `resized` signal to update its ellipsis. Here is when I encounter my main problem, the event is not called when I'm reducing the label size. This is because it has reached its `rect_min_size`, based on the currently shown text. The solution I came up with is the following:

```gdscript
func _label_width() -> int:
	# Without -1, the resulting rect_min_size would be equal to rect_size, which would not allow
	# resizing. With -1, rect_min_size is always at least 1 pixel less than rect_size, which allows
	# reducing its size.
	return int(get_size().x) - 1
```

I'm opening a discussion to know if I'm missing something. I'm not suggesting that this is a bug, I understand why `resized` is not called. Maybe the solution comes with better documentation on this subject.

**Comment by Calinou** [comment]
Godot 4.0alpha has built-in text ellipsis properties in Label. However, this can't be backported to `3.x` without a complete rewrite as it relies on backwards-incompatible TextServer changes.

**Reply by 98teg** [reply]
That's great! I would make sure to use it when I upgrade to Godot 4. But I think that my problem is a bit more general. I would summarize it as "being able to continue reducing the size of control nodes via code pass their minimum size". Here is another example, in the same app, I have created a grid layout that calculates the number of columns using its current size. It works great when I expand the size of the control node, adding columns as needed. But when I reduce its size, it reaches its minimum size and thus it doesn't call `resized` anymore. It thought that using [`_get_minimum_size`](https://docs.godotengine.org/en/stable/classes/class_control.html#class-control-method-get-minimum-size) would work, but as it says in the documentation, it ends up using the maximum values between this and `rect_min_size`.
